File: orbit-map/orbit_map/schemas/graph/nodes.py
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field

LeafKind = Literal[
    "function",
    "method",
    "class",
    "struct",
    "interface",
    "trait",
    "impl",
    "field",
    "module",
]


class BaseNode(BaseModel):
    id: str
    identity_key: str = ""
    object_hash: str | None = None
    name: str
    location: str
    language: str
    description: str = ""
    parent_id: str | None = None
    is_locked: bool = False
    lineage_locked: bool = False
    lock_owner: str | None = None
    lock_reason: str = ""


class SignatureField(BaseModel):
    name: str
    annotation: str | None = None
    description: str = ""


class LeafHistoryEntry(BaseModel):
    timestamp: datetime
    actor: str
    reason: str = ""
    source: str = ""
    source_blob_hash: str | None = None
    source_hash: str | None = None
    file_hash_at_capture: str | None = None


class DirNode(BaseNode):
    node_type: Literal["dir"] = "dir"
    dir_children: list[str] = Field(default_factory=list)
    file_children: list[str] = Field(default_factory=list)


class FileNode(BaseNode):
    node_type: Literal["file"] = "file"
    extension: str | None = None
    source_blob_hash: str | None = None
    imports: list[str] = Field(default_factory=list)
    exports: list[str] = Field(default_factory=list)
    leaf_children: list[str] = Field(default_factory=list)


class LeafNode(BaseNode):
    node_type: Literal["leaf"] = "leaf"
    kind: LeafKind
    source: str = ""
    source_blob_hash: str | None = None
    source_hash: str | None = None
    file_hash_at_capture: str | None = None
    history: list[LeafHistoryEntry] = Field(default_factory=list)
    input_signature: list[SignatureField] = Field(default_factory=list)
    output_signature: list[SignatureField] = Field(default_factory=list)
    start_line: int | None = None
    end_line: int | None = None
    children: list[str] = Field(default_factory=list)


class CodebaseGraphV1(BaseModel):
    root_dir_id: str
    dirs: list[DirNode] = Field(default_factory=list)
    files: list[FileNode] = Field(default_factory=list)
    leaves: list[LeafNode] = Field(default_factory=list)

File: crates/orbit-knowledge/src/graph/nodes.rs
use std::fmt;

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

// ---------------------------------------------------------------------------
// LeafKind
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum LeafKind {
    Function,
    FunctionDeclaration,
    Method,
    SingletonMethod,
    Class,
    SingletonClass,
    Package,
    Object,
    CompanionObject,
    Namespace,
    Enum,
    Struct,
    Record,
    Interface,
    TypeAlias,
    Trait,
    Impl,
    Property,
    Field,
    Event,
    Delegate,
    Global,
    Macro,
    Constant,
    Module,
    /// Markdown ATX heading (`#`–`######`). `depth` is 1–6. Added T20260422-1540.
    Section {
        depth: u8,
    },
    /// Top-level key in a structured config file (YAML / JSON / TOML).
    /// Added T20260422-1540.
    ConfigKey,
    /// Header cell in a tabular data file (CSV / TSV). Added T20260422-1540.
    Column,
}

impl fmt::Display for LeafKind {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let s = match self {
            Self::Function => "function",
            Self::FunctionDeclaration => "function_declaration",
            Self::Method => "method",
            Self::SingletonMethod => "singleton_method",
            Self::Class => "class",
            Self::SingletonClass => "singleton_class",
            Self::Package => "package",
            Self::Object => "object",
            Self::CompanionObject => "companion_object",
            Self::Namespace => "namespace",
            Self::Enum => "enum",
            Self::Struct => "struct",
            Self::Record => "record",
            Self::Interface => "interface",
            Self::TypeAlias => "type_alias",
            Self::Trait => "trait",
            Self::Impl => "impl",
            Self::Property => "property",
            Self::Field => "field",
            Self::Event => "event",
            Self::Delegate => "delegate",
            Self::Global => "global",
            Self::Macro => "macro",
            Self::Constant => "constant",
            Self::Module => "module",
            Self::Section { .. } => "section",
            Self::ConfigKey => "config_key",
            Self::Column => "column",
        };
        f.write_str(s)
    }
}

// ---------------------------------------------------------------------------
// SignatureField
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct SignatureField {
    pub name: String,
    #[serde(default)]
    pub annotation: Option<String>,
    #[serde(default)]
    pub description: String,
}

// ---------------------------------------------------------------------------
// LeafHistoryEntry
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct LeafHistoryEntry {
    pub timestamp: DateTime<Utc>,
    pub actor: String,
    #[serde(default)]
    pub reason: String,
    #[serde(default)]
    pub source: String,
    pub source_blob_hash: Option<String>,
    pub source_hash: Option<String>,
    pub file_hash_at_capture: Option<String>,
}

// ---------------------------------------------------------------------------
// BaseNodeFields
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct BaseNodeFields {
    pub id: String,
    #[serde(default)]
    pub identity_key: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub object_hash: Option<String>,
    pub name: String,
    pub location: String,
    pub language: String,
    #[serde(default)]
    pub description: String,
    pub parent_id: Option<String>,
    #[serde(default)]
    pub is_locked: bool,
    #[serde(default)]
    pub lineage_locked: bool,
    pub lock_owner: Option<String>,
    #[serde(default)]
    pub lock_reason: String,
}

// ---------------------------------------------------------------------------
// DirNode
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct DirNode {
    #[serde(flatten)]
    pub base: BaseNodeFields,
    #[serde(default)]
    pub dir_children: Vec<String>,
    #[serde(default)]
    pub file_children: Vec<String>,
}

// ---------------------------------------------------------------------------
// FileNode
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct FileNode {
    #[serde(flatten)]
    pub base: BaseNodeFields,
    pub extension: Option<String>,
    pub source_blob_hash: Option<String>,
    #[serde(default)]
    pub source: String,
    #[serde(default)]
    pub imports: Vec<String>,
    #[serde(default)]
    pub exports: Vec<String>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub re_exports: Vec<ReExport>,
    #[serde(default)]
    pub leaf_children: Vec<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct ReExport {
    pub name: String,
    pub source_path: String,
}

// ---------------------------------------------------------------------------
// LeafNode
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct LeafNode {
    #[serde(flatten)]
    pub base: BaseNodeFields,
    pub kind: LeafKind,
    #[serde(default)]
    pub source: String,
    pub source_blob_hash: Option<String>,
    pub source_hash: Option<String>,
    pub file_hash_at_capture: Option<String>,
    #[serde(default)]
    pub history: Vec<LeafHistoryEntry>,
    #[serde(default)]
    pub input_signature: Vec<SignatureField>,
    #[serde(default)]
    pub output_signature: Vec<SignatureField>,
    pub start_line: Option<u32>,
    pub end_line: Option<u32>,
    #[serde(default)]
    pub children: Vec<String>,
}

// ---------------------------------------------------------------------------
// GraphNode (untagged wrapper — constructed programmatically, not via serde)
// ---------------------------------------------------------------------------

#[derive(Debug, Clone)]
pub enum GraphNode {
    Dir(DirNode),
    File(FileNode),
    Leaf(LeafNode),
}

impl GraphNode {
    pub fn id(&self) -> &str {
        &self.base().id
    }

    pub fn base(&self) -> &BaseNodeFields {
        match self {
            Self::Dir(n) => &n.base,
            Self::File(n) => &n.base,
            Self::Leaf(n) => &n.base,
        }
    }

    pub fn location(&self) -> &str {
        &self.base().location
    }

    pub fn parent_id(&self) -> Option<&str> {
        self.base().parent_id.as_deref()
    }
}

// ---------------------------------------------------------------------------
// CodebaseGraphV1
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CodebaseGraphV1 {
    pub root_dir_id: String,
    #[serde(default)]
    pub dirs: Vec<DirNode>,
    #[serde(default)]
    pub files: Vec<FileNode>,
    #[serde(default)]
    pub leaves: Vec<LeafNode>,
}

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------
